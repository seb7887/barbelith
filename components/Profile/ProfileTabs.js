import React from 'react';
import AppBar from '@material-ui/core/AppBar';
import Typography from '@material-ui/core/Typography';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';

import Post from '../Home/Post';
import FollowTab from './FollowTab';

const TabContainer = ({ children }) => (
  <Typography component='div' style={{ padding: '1em' }}>
    {children}
  </Typography>
)

class ProfileTabs extends React.Component {
  state = {
    tab: 0
  };

  handleTabChange = (e, value) => {
    e.preventDefault();
    this.setState({ tab: value });
  }

  render() {
    const { tab } = this.state;
    const {
      auth,
      posts,
      user,
      isDeletingPost,
      handleDeletePost,
      handleToggleLike,
      handleAddComment,
      handleDeleteComment
    } = this.props;

    return (
      <div>
        <AppBar position='static' color='default'>
          <Tabs
            value={tab}
            onChange={this.handleTabChange}
            indicatorColor='secondary'
            textColor='secondary'
            variant='fullWidth'
          >
            <Tab label='Posts' data-testid='tab1' />
            <Tab label='Following' data-testid='tab2' />
            <Tab label='Followers' data-testid='tab3' />
          </Tabs>
        </AppBar>
        {tab === 0 && (
          <TabContainer>
            {posts.map(post => (
              <Post
                key={post._id}
                auth={auth}
                post={post}
                isDeletingPost={isDeletingPost}
                handleDeletePost={handleDeletePost}
                handleToggleLike={handleToggleLike}
                handleAddComment={handleAddComment}
                handleDeleteComment={handleDeleteComment}
              />
            ))}
          </TabContainer>
        )}
        {tab === 1 && (
          <TabContainer>
            <FollowTab users={user.following} />
          </TabContainer>
        )}
        {tab === 2 && (
          <TabContainer>
            <FollowTab users={user.followers} />
          </TabContainer>
        )}
      </div>
    );
  }
}

export default ProfileTabs;