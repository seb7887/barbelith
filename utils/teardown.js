use barbelith-db;
show collections;
db.users.remove({});
db.sessions.remove({});
db.posts.remove({});
