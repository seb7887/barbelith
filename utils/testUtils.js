import faker from 'faker';

export const validUser = () => ({
  name: 'kingmob',
  email: 'king@mob.com',
  password: 'km1234'
});

export const fakeUser = () => ({
  name: faker.internet.userName(),
  email: faker.internet.email(),
  password: faker.internet.password()
});

export const mockedUser = () => ({
  _id: '7',
  avatar: '/static/images/avatar.jpg',
  createdAt: '2019-02-1234',
  email: faker.internet.email(),
  followers: ['1', '2', '3'],
  following: ['3'],
  name: 'Test',
  updatedAt: '2019-02-1234'
});

export const updatedUser = () => ({
  _id: '7',
  name: 'kingmob',
  email: faker.internet.email(),
  avatar: faker.image.avatar(),
  about: faker.lorem.text()
});

export const generateUsers = quantity => {
  const generatedUsers = [];

  for (let i = 0; i < quantity; i++) {
    const user = {
      _id: `${i}`,
      name: `test${i}`,
      avatar: faker.image.avatar()
    };
    generatedUsers.push(user);
  }

  return generatedUsers;
};

export const generateComments = quantity => {
  const generatedComments = [];

  for (let i = 0; i < quantity; i++) {
    const comment = {
      _id: faker.random.uuid().toString(),
      postedBy: {
        _id: '7',
        name: 'test',
        avatar: faker.image.avatar()
      },
      text: faker.lorem.text()
    };
    generatedComments.push(comment);
  }
  return generatedComments;
};

export const fakePost = () => ({
  _id: '7',
  postedBy: {
    _id: '7',
    name: 'test',
    avatar: faker.image.avatar()
  },
  createdAt: faker.date.recent().toString(),
  text: faker.lorem.text(),
  image: faker.image.avatar(),
  likes: ['7'],
  comments: []
});
