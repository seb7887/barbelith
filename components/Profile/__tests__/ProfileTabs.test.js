/* eslint-env jest */
import 'react-testing-library/cleanup-after-each';

import { render, fireEvent, waitForElement } from 'react-testing-library';
import { fakePost, fakeUser } from '../../../utils/testUtils';
import ProfileTabs from '../ProfileTabs';

let auth = {
  user: {
    _id: '7'
  }
};

const user = fakeUser();
user.following = [];
user.following.push(fakeUser());
user.following[0]._id = '8';
const post = fakePost();

const handleDeletePost = jest.fn();
const handleToggleLike = jest.fn();
const handleAddComment = jest.fn();
const handleDeleteComment = jest.fn();

jest.mock('next/link', () => {
  return ({ children }) => {
    return children;
  }
});

describe('<ProfileTabs/>', () => {
  it('renders and matches snapshot', () => {
    const { asFragment } = render(
      <ProfileTabs
        auth={auth}
        posts={[]}
        user={{}}
        isDeletingPost={false}
        handleDeletePost={handleDeletePost}
        handleToggleLike={handleToggleLike}
        handleAddComment={handleAddComment}
        handleDeleteComment={handleDeleteComment}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('changes tab correctly', async () => {
    const { getByTestId } = render(
      <ProfileTabs
        auth={auth}
        posts={[post]}
        user={user}
        isDeletingPost={false}
        handleDeletePost={handleDeletePost}
        handleToggleLike={handleToggleLike}
        handleAddComment={handleAddComment}
        handleDeleteComment={handleDeleteComment}
      />
    );

    const content = await waitForElement(() => getByTestId('content'));
    expect(content.textContent).toContain(post.text);

    fireEvent.click(getByTestId('tab2'));
    const following = await waitForElement(() => getByTestId('follow-tab'));
    expect(following.textContent).toContain(user.following[0].name);
  });
});
