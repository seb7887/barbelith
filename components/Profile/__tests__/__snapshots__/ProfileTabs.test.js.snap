// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileTabs/> renders and matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <header
      class="MuiPaper-root-10 MuiPaper-elevation4-16 MuiAppBar-root-1 MuiAppBar-positionStatic-5 MuiAppBar-colorDefault-7"
    >
      <div
        class="MuiTabs-root-37"
      >
        <div
          class="MuiTabs-flexContainer-38"
        >
          <div
            class="MuiTabs-scroller-40 MuiTabs-fixed-41"
            role="tablist"
            style="margin-bottom: 0px;"
          >
            <div
              class="MuiTabs-flexContainer-38"
            >
              <button
                aria-selected="true"
                class="MuiButtonBase-root-58 MuiTab-root-46 MuiTab-textColorSecondary-50 MuiTab-selected-51 MuiTab-fullWidth-53"
                data-testid="tab1"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper-54"
                >
                  <span
                    class="MuiTab-labelContainer-55"
                  >
                    <span
                      class="MuiTab-label-56"
                    >
                      Posts
                    </span>
                  </span>
                </span>
                <span
                  class="MuiTouchRipple-root-100"
                />
              </button>
              <button
                aria-selected="false"
                class="MuiButtonBase-root-58 MuiTab-root-46 MuiTab-textColorSecondary-50 MuiTab-fullWidth-53"
                data-testid="tab2"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper-54"
                >
                  <span
                    class="MuiTab-labelContainer-55"
                  >
                    <span
                      class="MuiTab-label-56"
                    >
                      Following
                    </span>
                  </span>
                </span>
                <span
                  class="MuiTouchRipple-root-100"
                />
              </button>
              <button
                aria-selected="false"
                class="MuiButtonBase-root-58 MuiTab-root-46 MuiTab-textColorSecondary-50 MuiTab-fullWidth-53"
                data-testid="tab3"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper-54"
                >
                  <span
                    class="MuiTab-labelContainer-55"
                  >
                    <span
                      class="MuiTab-label-56"
                    >
                      Followers
                    </span>
                  </span>
                </span>
                <span
                  class="MuiTouchRipple-root-100"
                />
              </button>
            </div>
            <span
              class="MuiPrivateTabIndicator-root-61 MuiPrivateTabIndicator-colorSecondary-63 MuiTabs-indicator-45"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
    </header>
    <div
      class="MuiTypography-root-64 MuiTypography-body1-73"
      style="padding: 1em;"
    />
  </div>
</DocumentFragment>
`;
